---
title: "Master"
author: "Maja Kuzman"
date: "November 16, 2018"
output: html_notebook
---


This notebook reproduces all figures made by custom R code produced by bioinformatics group of University of Zagreb. It will save them in the files with appropriate names.  
To knit this markdown document you will need R version>=3.5, and additional packages that will now be installed:

```{r, message=FALSE}
check.packages <- function(pkg){
     if (!requireNamespace("BiocManager", quietly = TRUE))
          install.packages("BiocManager")
     new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
     if (length(new.pkg))
         BiocManager::install(new.pkg)
     new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
     if (length(new.pkg)) 
        install.packages(new.pkg, dependencies = TRUE)
     sapply(pkg, require, character.only = TRUE)
}
check.packages(c("GenomicRanges", "data.table","digest", "ggplot2", "parallel", "stringr", "pheatmap", "RColorBrewer"))
```


# Figure 1.  
## HIV-1 integration hotspots are within genes delineated with super-enhancers.  
#### A) Metagene plots of H3K27ac, H3K4me1, H3K4me3, BRD4, MED1, H3K36me3, H4K20me1, H3K9me2 and H3K27me3 ChIP-Seq signals in recurrent integration genes (RIGs) in red and the rest of the genes that are not targeted by HIV-1 (no RIGs) in black.  

Figure was generated using ngsplot software.  

#### B) ROC analysis represented in heatmap summarizing the co-occurrence of integration sites and epigenetic modification obtained by ChIP-Seq for H3K27ac, H3K4me1, BRD4, MED1, H3K36me3, H4K20me1, H3K4me3, H3K27me3 and H3K9me2. HIV-1 integration data sets are shown in the columns, and epigenetic modifications are shown in rows. Associations are quantified using the ROC area method; values of ROC areas are shown in the color key at the right.  

Figures 1C and Supplementary 1E were generated using custom script below.  

The script generates 4 other documents:   

1. ROC_SupplementaryFigure_byBinsizes is the document associated with supplementary figure 1E. ROC analysis represented as heatmap summarizing the co-occurrence density of integration sites and epigenetic modification obtained by ChIP-Seq for H3K27ac, H3K4me1, BRD4, MED1, H3K36me3, H4K20me1, H3K4me3, H3K27me3 and H3K9me2. HTLV, HIV-1 and MLV integration data sets are shown in the columns, and epigenetic modifications are shown in rows. Associations are quantified using the ROC area method; values of ROC areas are shown in the color key at the right. The figure shows results for all bin sizes used, while in paper only bin size of 1000 is shown.  

2. ROC_allValues: is the document which contains co-occurence densities for all bin sizes and for all epigenetic modifications, for all analyzed data sets. Numbers represent calculated phi values.  

3. ROC_mainFigure_byBinsizes: ROC values for all bins, in the paper we have shown values for bin size=1000, as figure 1B.  

4. ROC_withasterisks: significance levels are indicated with asterisks, pval<0.05 is indicated with a single *, p<0.01 with **,
and p < 0.001 as ***  

```{r, fig.height=15, fig.width=7}
source("ROC_figures.R")

```


#### C) Distance to the nearest super-enhancer in activated CD4+ T cells. Box plots represent distance from the gene to nearest super-enhancer (SE) grouped by number of times gene is found in different data sets.  

Figure was generated using custom script (AllFigures.R):  
```{r}
source("AllFigures.R")
Fig1C
```


#### D) FOXP1 , STAT5B and BACH2 ALL IS (black) superimposition on H3K27Ac (orange), SE (blue), H3K36me3 (green) and BRD4 (violet) ChIP-Seq tracks.   

Figure was generated by plotting ChIP signals on genes.

# Figure 2.  
## RIGs are delineated with super-enhancers regardless of their expression.  

#### A) Regularized log transformed read counts on genes averaged over three replicates in activated CD4+ T cells shown as violin plot for genes without HIV-1 integrations and genes with HIV-1 integrations.  

Figure was generated using custom script (AllFigures.R):  
```{r, message=FALSE}
Fig2A
```

#### B) Box plot for genes grouped by number of HIV-1 lists they appear in.  

Figure was generated using custom script (AllFigures.R):  
```{r, message=FALSE}
Fig2B     
```

#### C) Box plot for genes grouped by number of HIV-1 lists they appear in, with RIGs grouped together in 2 or more lists group. Box plots are shown separately for genes that have super-enhancer 5 kb upstream of TSS or super-enhancer overlaps them (SE in proximity), and genes that do not have super-enhancer in proximity. Differences in median abundances of mRNA are statistically significant for all groups (p-value < 2.2x10 -16 for genes without HIV integrations and genes found on only one list and p-value 3.7x10-12 for RIGs, calculated by Wilcoxon rank sum test).   

Figure was generated using custom script (AllFigures.R):  
```{r, message=FALSE}
Fig2C
```

#### D) Bar plot shows percentage of genes that have super-enhancer in proximity, divided by number of lists gene is found in and expression group.  

Figure was generated using custom script AllFigures.R:  
```{r}
Figure2D

```

# Figure 3.  
## Super-enhancer unloading in activated CD4+ T cells does not impact HIV-1 integration patterns.  
#### D) Regularized log transformed read counts on genes averaged over three replicates in activated JQ1 treated cells shown as violin plot for genes grouped by presence of HIV-1 integration in activated JQ1 treated cells.  

Figure was generated using custom script AllFigures.R:

```{r}
Fig3D
```

# Supplementary figure 1.  

#### A) HIV-1 integration sites inside genes. The box plot represents the percentage of integration sites inside genes for cART-treated patients in violet (6 lists in total) and in vitro infections in blue (2 lists in total). The red line depicts the median and whiskers stretch from 5th to 95th percentile.  

```{r}
source("Supplementary figure 1a.R")
supplementary1a
```


#### B) Analysis of the number of unique genes containing integration sites versus number of observed integrations. All integration datasets are sorted by decreasing size and the cumulative number of integration sites is plotted on the X-axis while the Y-axis shows the number of unique genes that have integrations. The number of unique genes found when analyzing different numbers of datasets linearly depends on the number of integrations in the observed data sets (adjusted R2 = 0.9849, p = 7.7x10-8).  

Plot is generated using a custom script:
```{r}
source("Saturation curve.R")
```

#### C) HIV-1 recurrently integrates into a subset of genes. The bar plot represents the number of genes (RIGs) shared among at least x different data sets. The number of RIGs shared among different data sets decreases exponentially as more data sets are taken into consideration.  

Plot is made in excel from table available as Aditional file 2.

#### D) Bar plot showing the percentage of genes that have a super-enhancer in proximity, in groups of genes on 0 list (without HIV-1 integrations), genes on 1 list and genes in 2 or more lists (RIGs).  

Plot is done with custom script AllFigures.R:
```{r}
SupplementaryFig1D
```


#### E) ROC analysis represented as heatmap summarizing the co-occurrence density of integration sites and epigenetic modification obtained by ChIP-Seq for H3K27ac, H3K4me1, BRD4, MED1, H3K36me3, H4K20me1, H3K4me3, H3K27me3 and H3K9me2. HTLV, HIV-1 and MLV integration data sets are shown in the columns, and epigenetic modifications are shown in rows. Associations are quantified using the ROC area method; values of ROC areas are shown in the color key at the right.  

Plotting is already demonstrated, see Figure 1B and ROC_figures.R, and file SupplementaryFigure1E.pdf

# Supplementary figure 2:  
#### B) Volcano plot showing the changes in mRNA levels of genes upon JQ1 treatment with respect to the vicinity to super-enhancers.  
Plot is done with custom script AllFigures.R:

```{r, warning=FALSE}
SupplementaryFig2B
```


#### C) Bar plot showing the percentage of genes that are downregulated, unchanged, and upregulated upon JQ1 treatment. Genes are grouped by number of lists they occur in and by the presence or absence of super-enhancer in either gene body or 5 kb upstream.  

Plot is done with custom script AllFigures.R:

```{r}
SupplementaryFig2C
```






# Supplementary figure 3:  

Supplementary figure 3 is produced with the following custom code:  

```{r}
source("Supplementary figure 3.R")
```

